### PR TITLE
`type_of` metafunction should dealias the type

### LINKED ISSUE #91: `type_of` metafunction should dealias the type
P2996R6 will change `type_of` to no longer return reflections representing `$typedef-name$`s - instead, a reflection of the underlying type will be returned (see [here](https://godbolt.org/z/eqhGTWMda)).

The `type_of` function is implemented [here](https://github.com/bloomberg/clang-p2996/blob/e97e8c2eeab71377c4b1a5cbd54deb3c9a020c1e/clang/lib/Sema/Metafunctions.cpp#L2000-L2054), and we should also update the `return_type_of` metafunction (proposed by P3096), implemented [here](https://github.com/bloomberg/clang-p2996/blob/e97e8c2eeab71377c4b1a5cbd54deb3c9a020c1e/clang/lib/Sema/Metafunctions.cpp#L5557-L5590).

We have an [existing `desugarType` utility](https://github.com/bloomberg/clang-p2996/blob/e97e8c2eeab71377c4b1a5cbd54deb3c9a020c1e/clang/lib/Sema/Metafunctions.cpp#L839-L874) in `Metafunctions.cpp` that often helps with tasks like these - there is an `UnwrapAliases` argument that should be set to `true`.

[libcxx reflection tests](https://github.com/bloomberg/clang-p2996/tree/p2996/libcxx/test/std/experimental/reflection) should of course be updated to reflect this change.

### PR DESCRIPTION
**Pull request by changkhothuychung**
Fix #91

*Issue number of the reported bug or feature request: #<number>*

**Describe your changes**
Adding code to dealias the underlying type in case of using typedef-name

**Testing performed**
Updating current tests in libcxx to reflect the new expected behaviors. 

**Additional context**

### CODE REVIEW

**Comment by changkhothuychung** [comment]
Hi @katzdm , can you help me take a look when you have time?

File: clang/lib/Sema/Metafunctions.cpp [hunk 1]
```diff
@@ -2007,7 +2007,15 @@ bool type_of(APValue &Result, Sema &S, EvalFn Evaluator, DiagFn Diagnoser,
 
   switch (RV.getReflectionKind()) {
   case ReflectionKind::Null:
-  case ReflectionKind::Type:
```
Comment (katzdm) [inline clang/lib/Sema/Metafunctions.cpp]: I probably should have pointed to this earlier, but this is the spec for `type_of`:
https://brevzin.github.io/cpp_proposals/2996_reflection/d2996r6.html#pnum_210

`type_of(^^int)` fails to be a constant expression, so the existing behavior (i.e., diagnosing an error) is correct.

The cases that need modification are `Object`, `Value`, `Declaration`, `BaseSpecifier`, and `DataMemberSpec`.

(also for `return_type_of` below)

Ping me tomorrow if there's anything I can help with!

File: clang/lib/Sema/Metafunctions.cpp [hunk 2]
```diff
@@ -2007,7 +2007,15 @@ bool type_of(APValue &Result, Sema &S, EvalFn Evaluator, DiagFn Diagnoser,
...
+  case ReflectionKind::Type: {
+    QualType currentType = RV.getTypeOfReflectedResult(S.Context);
+    bool UnwrapAlises = isa<UsingType>(currentType) ||
+                        isa<TypedefType>(currentType) ||
+                        isa<TemplateSpecializationType>(currentType);
```
Comment (katzdm) [inline clang/lib/Sema/Metafunctions.cpp]: Out of curiosity, what happens if we just pass `/*UnwrapAliases=*/true` unconditionally? Are we avoiding an undesirable behavior with this check, or is this an optimization? If the later, I'd prefer passing unconditional `true`.

(also for `return_type_of` below)

Comment (changkhothuychung) [reply clang/lib/Sema/Metafunctions.cpp]: I updated to unconditional true and it looks fine to me. I updated some tests in libcxx to reflect the new behavior.

**Comment by changkhothuychung** [comment]
I have this test fails locally - `std/experimental/reflection/module-imports.sh.cpp`

The reason is probably due to the arch of my laptop (mac M1). From the log - ` #  error "We don't know how to get the definition of mbstate_t without <wchar.h> on your platform."`

Can you help me run that test with my change to see if that runs ok?

**Comment by katzdm** [comment]
> I have this test fails locally - `std/experimental/reflection/module-imports.sh.cpp`
> 
> The reason is probably due to the arch of my laptop (mac M1). From the log - ` # error "We don't know how to get the definition of mbstate_t without <wchar.h> on your platform."`
> 
> Can you help me run that test with my change to see if that runs ok?

@changkhothuychung I believe I fixed this yesterday evening, while merging latest changes from upstream. Let me know if pulling from latest fixes the issue for you.

**Comment by changkhothuychung** [comment]
@katzdm I pulled the latest from p2996, but it doesnt seem to fix the error above.

**Review by katzdm — APPROVED** [review]

**Pull request merged**